### ISSUE TITLE
Question regarding Number of components to select

### ISSUE THREAD

**Issue by amnahsiddiqa** [BiRG/pyopls#4]
Hi I have a quick question;I will appreciate if you could explain that;
I see that in OPLS 39 components are used and in pls one component is passed in the code below;
#opls = OPLS(39)
#Z = opls.fit_transform(spectra, target)
#pls = PLSRegression(1)

Q1: What is the reason for using one component in "PLSRegression(1)" and how do you decide the number of OPLS components  OPLS(39);

Q2: In Permutation test; what is this argument k=-1 referring to? and where can we access the  information about what numbers of components were best for this model after doing permutation test? 
#validator = OPLSValidator(k=-1).fit(X, target)

I appreciate you time and response;
Thanks
Amnah

**Comment by dpfoose** [comment]
@amnahsiddiqa 
Hello,
Q1: we only need one component in the PLSRegression classifier because we've removed all information that is orthogonal to the target. Using additional PLS components will not improve classifier performance if the optimal number of OPLS components are selected. 39 components are chosen in the example because this was the number of components that maximizes _Q<sup>2</sup>_ (the _R<sup>2</sup>_ value for the left-out samples in the leave-one-out CV). This is found using the OPLSValidator class. It is possible to use OPLS as preprocessing for other estimators such as support-vector machines (which has been tried in the metabolomics literature), but the 1-component PLS estimator was used in the original Wold paper, and the choice of downstream estimator and metric is not that important for the determination of the optimal number of OPLS components.

Q2: You choose k=-1 in the OPLSValidator class to select leave-one-out cross validation instead of k-fold. This is similar to the convention used by scikit-learn for `n_jobs` in parallel operations, where -1 uses the maximum number of processor cores, or in the Python indexing operator, where the -1<sup>th</sup> element is the last element in the list.

`OPLSValidator` will try every number of OPLS components in up to (n-1) until the _Q<sup>2</sup>_ for the 1-component PLS model no longer increases (this is possible because _Q<sup>2</sup>_ and other cross-validated metrics, unlike _R<sup>2</sup>_, will tend to decrease as the model becomes overfit, so there exists an optimal value for the number of OPLS components _p_ which maximizes _Q<sup>2</sup>_). The goal of OPLS is to remove orthogonal components from the original data so as to improve the PLS model's performance. We are choosing _p_ for our OPLS pre-processing such that performance of the downstream PLS estimator is optimized.
